include/nbl/builtin/glsl/ext/RadixSort/default_histogram.comp

```diff
@@ -0,0 +1,124 @@
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_BINDING_DEFINED_ 1
+#endif
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_INOUT_DESCRIPTOR_DEFINED_
+
+layout (set = _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_INOUT_SET_DEFINED_, binding = _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_INOUT_BINDING_DEFINED_, std430) buffer inout_buffer
+{
+	uvec2 inout_values[];
+};
+
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_INOUT_DESCRIPTOR_DEFINED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_DESCRIPTOR_DEFINED_
+
+layout (set = _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_SET_DEFINED_, binding = _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_BINDING_DEFINED_, std430) writeonly buffer histogram_out_buffer
+{
+	uint histogram_out_values[];
+};
+
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_OUTPUT_DESCRIPTOR_DEFINED_
+#endif
+
+#include "nbl/builtin/glsl/ext/RadixSort/parameters_struct.glsl"
+#include "nbl/builtin/glsl/ext/RadixSort/parameters.glsl"
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_PUSH_CONSTANTS_DEFINED_
+
+layout(push_constant) uniform PushConstants
+{
+	layout (offset = 0) nbl_glsl_ext_RadixSort_Parameters_t params;
+} pc;
+
+#define _NBL_GLSL_EXT_RADIXSORT_PUSH_CONSTANTS_DEFINED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_GET_PARAMETERS_DEFINED_
+
+nbl_glsl_ext_RadixSort_Parameters_t nbl_glsl_ext_RadixSort_getParameters()
+{
+	return pc.params;
+}
+
+#define _NBL_GLSL_EXT_RADIXSORT_GET_PARAMETERS_DEFINED_
+#endif
+
+#include <nbl/builtin/glsl/limits/numeric.glsl>
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_GET_PADDED_DATA_DEFINED_
+
+uvec2 nbl_glsl_ext_RadixSort_getPaddedData(in uint idx)
+{
+	uvec2 data = uvec2(UINT_MAX);
+	if (idx < nbl_glsl_ext_RadixSort_Parameters_t_getElementCountTotal())
+		data = inout_values[idx];
+
+	return data;
+}
+
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_GET_PADDED_DATA_DEFINED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_SET_DATA_DEFINED_
+
+void nbl_glsl_ext_RadixSort_setData(in uint idx, in uvec2 data)
+{
+	if (idx < nbl_glsl_ext_RadixSort_Parameters_t_getElementCountTotal())
+		inout_values[idx] = data;
+}
+
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_SET_DATA_DEFINED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_SET_HISTOGRAM_DEFINED_
+
+void nbl_glsl_ext_RadixSort_setHistogram(in uint idx, in uint val)
+{
+	histogram_out_values[idx] = val;
+}
+
+#define _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_SET_HISTOGRAM_DEFINED_
+#endif
+
+#include "nbl/builtin/glsl/ext/RadixSort/radix_sort.glsl"
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_HISTOGRAM_MAIN_DEFINED_
+
+void main()
+{
+	uvec2 data = nbl_glsl_ext_RadixSort_getPaddedData(gl_GlobalInvocationID.x);
+	uint digit = nbl_glsl_ext_RadixSort_extractDigit(data.x);
+
+	uint scatter_idx = nbl_glsl_ext_RadixSort_workgroupCompact(digit, true);
+
+	if (gl_LocalInvocationIndex == 0u)
+	{
+		for (int i = 0; i < NUM_BUCKETS; ++i)
+			nbl_glsl_ext_RadixSort_setHistogram(i * gl_NumWorkGroups.x + gl_WorkGroupID.x, scratch_shared[i]);
+	}
```

Comment:
this is weird an ugly, the last invocation has the `local_histogram` in registers not shared memory, and you needlessly force a barrier and smem transfer to the first invocation so the first invocation can write the histogram.

I'd rather `nbl_glsl_ext_RadixSort_workgroupCompact` call `nbl_glsl_ext_RadixSort_setHistogram` inside th branch `if (save_local_histogram && (gl_GlobalInvocationID.x == last_of_wg_idx))` then you can have one `barrier()` less

Reply:
> I'd rather nbl_glsl_ext_RadixSort_workgroupCompact call nbl_glsl_ext_RadixSort_setHistogram inside th branch if (save_local_histogram && (gl_GlobalInvocationID.x == last_of_wg_idx)) then you can have one barrier() less

This was _exactly_ what was being done before. I had to split it up the entire thing into two functions (compact and setHistogram) because there was a possibility of (glsl) linker error if `radix_sort.glsl` wasn't used correctly.

Notice, `default_histogram.comp` can write to the histogram descriptor while `default_scatter.comp` cannot (`readonly`) or rather it wouldn't make sense to write to histogram here. If I do what you suggest, I'd have to put `nbl_glsl_ext_RadixSort_setHistogram` in `radix_sort.glsl` (the main algorithm header), which is going to be included in both `default_histogram.comp` and `default_scatter.comp`, that means you either need to provide a dummy implementation of `nbl_glsl_ext_RadixSort_setHistogram` in `default_scatter.comp` or you'd have to risk a linker error resulting from potential incorrect use of `radix_sort.glsl` by removing the include guard (in `radix_sort.glsl`) which checks if the set histogram method is defined.

This is one of those problems resulting from not knowing how the API will actually be used.

Reply:
Maybe make two different histogram functions , or rather split this into the histogram and then histogram saving function?

Reply:
what about splitting `radix_sort.glsl` to `radix_sort_histogram.glsl` and `radix_sort_scatter.glsl`?

Reply:
decent idea

Reply:
hey i have another idea

can i make it optionally compilable like,

```
uint nbl_glsl_ext_RadixSort_workgroupCompact(in uint digit)
{
	uint scatter_idx = 0u;
	uint local_histogram[_NBL_GLSL_EXT_RADIXSORT_BUCKET_COUNT_];
	for (int i = 0; i < _NBL_GLSL_EXT_RADIXSORT_BUCKET_COUNT_; ++i)
	{
		const bool predicate = (i == digit);
		nbl_glsl_workgroupBallot(predicate);
		local_histogram[i] = nbl_glsl_workgroupBallotInclusiveBitCount();

		if (predicate)
			scatter_idx = local_histogram[i];
	}

#ifdef _NBL_GLSL_EXT_RADIXSORT_SET_DATA_DEFINED_
	uint last_of_wg_idx = min((gl_WorkGroupID.x + 1u) * _NBL_GLSL_WORKGROUP_SIZE_ - 1u, nbl_glsl_ext_RadixSort_Parameters_t_getElementCountTotal() - 1u);
	if (gl_GlobalInvocationID.x == last_of_wg_idx)
	{
		for (int i = 0; i < _NBL_GLSL_EXT_RADIXSORT_BUCKET_COUNT_; ++i)
			nbl_glsl_ext_RadixSort_setHistogram(i * gl_NumWorkGroups.x + glWorkGroupID.x, local_histogram[i]);
	}
#endif
	return scatter_idx - 1u;
}
```
this way anyone who wants to write to the global histogram (for example `default_histogram.comp` and things derived from it, or some completely different shader) will set the macro in question and get the relevant code compiled while others (like `default_scatter.comp`) won't define the macro and doesn't compile the code which at the risk of linker error